### PR TITLE
Listing all available databases in the _ingest/geoip/database API

### PR DESCRIPTION
**Pull request by masseyke**
This returns all available databases from _ingest/geoip/database, rather than just the configured maxmind databases. Previously, all databases returned had a `maxmind` field, because they were only the databases that had been created using the `/_ingest/geoip/database` API. Now the API will also return databases downloaded by the standard downloader (with type `web`) and databases that were manually put into the config/ingest-geoip directory on any node (with type `local`). These `web` and `local` databases are read-only, and cannot be updated or deleted through the API. Here's an example API response:
```
$ curl "localhost:9200/_ingest/geoip/database?pretty&human=true"
{
  "databases" : [
    {
      "id" : "_web_R2VvTGl0ZTItQVNOLm1tZGI=",
      "version" : -1,
      "modified_date" : "2024-09-25T19:23:15.521Z",
      "modified_date_millis" : 1727292195521,
      "database" : {
        "name" : "GeoLite2-ASN",
        "web" : { }
      }
    },
    {
      "id" : "_web_R2VvTGl0ZTItQ291bnRyeS5tbWRi",
      "version" : -1,
      "modified_date" : "2024-09-25T19:23:23.900Z",
      "modified_date_millis" : 1727292203900,
      "database" : {
        "name" : "GeoLite2-Country",
        "web" : { }
      }
    },
    {
      "id" : "_web_R2VvTGl0ZTItQ2l0eS5tbWRi",
      "version" : -1,
      "modified_date" : "2024-09-25T19:23:18.109Z",
      "modified_date_millis" : 1727292198109,
      "database" : {
        "name" : "GeoLite2-City",
        "web" : { }
      }
    },
    {
      "id" : "geoip-city-enterprise",
      "version" : 1,
      "modified_date" : "2024-09-26T14:05:50.138Z",
      "modified_date_millis" : 1727359550138,
      "database" : {
        "name" : "GeoIP2-City",
        "maxmind" : {
          "account_id" : "1234"
        }
      }
    },
    {
      "id" : "_local_R2VvSVAyLUNpdHktVGVzdC5tbWRi",
      "version" : -1,
      "modified_date" : "2024-07-24T16:29:44.000Z",
      "modified_date_millis" : 1721838584000,
      "database" : {
        "name" : "GeoIP2-City-Test",
        "local" : {
          "type" : "GeoIP2-City"
        }
      }
    }
  ]
}
```

### CODE REVIEW

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java [hunk 1]
```diff
@@ -173,15 +173,17 @@ void updateDatabases() throws IOException {
                 final String id = entry.getKey();
                 DatabaseConfiguration database = entry.getValue().database();
                 if (existingDatabaseNames.contains(database.name() + ".mmdb") == false) {
-                    logger.debug("A new database appeared [{}]", database.name());
+                    if (database.provider() instanceof DatabaseConfiguration.Maxmind maxmind) {
+                        logger.debug("A new database appeared [{}]", database.name());
+                        final String accountId = maxmind.accountId();
+                        try (HttpClient.PasswordAuthenticationHolder holder = credentialsBuilder.apply(accountId)) {
+                            if (holder == null) {
```
Comment (lukewhiting) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:180]: We're getting quite a few layers deep here. Is it worth extracting some of these blocks to methods for readability?

File: modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfigurationTests.java [hunk 2]
```diff
@@ -35,26 +44,39 @@ protected DatabaseConfiguration createTestInstance() {
...
                 );
             case 2:
-                return new DatabaseConfiguration(
-                    instance.id(),
-                    instance.name(),
-                    new Maxmind(instance.maxmind().accountId() + randomAlphaOfLength(2))
-                );
+                DatabaseConfiguration.Provider provider = instance.provider();
+                DatabaseConfiguration.Provider modifiedProvider;
+                if (provider instanceof Maxmind maxmind) {
```
Comment (lukewhiting) [inline modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfigurationTests.java:70]: Would switch pattern matching be cleaner here? I think we are targeting > Java 21?

**Review by lukewhiting — COMMENTED** [review]
I can't give a definitive approval as I'm not familiar enough with the functionality yet but bar some minor readability nitpicks, code structure looks good, tests seem to cover the changes and implementation makes sense in context of the PR's description.

File: modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfigurationTests.java (hunk 2 above)
Comment (masseyke) [reply modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfigurationTests.java:70]: We've got to backport this to 8.x, which is on Java 17. I don't think switching by type is supported in 17 (I'll double-check).

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java (hunk 1 above)
Comment (masseyke) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java]: You caught us in the middle of a few changes ::) -- @joegallo took care of this in another PR that I just merged into this one.

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfiguration.java [hunk 3]
```diff
@@ -105,14 +132,24 @@ public static DatabaseConfiguration parse(XContentParser parser, String id) {
...
         out.writeString(id);
         out.writeString(name);
-        maxmind.writeTo(out);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.INGEST_GEO_DATABASE_PROVIDERS)) {
+            out.writeNamedWriteable(provider);
+        } else {
+            if (provider instanceof Maxmind maxmind) {
+                out.writeString(maxmind.accountId);
+            } else {
+                // the existence of a non-Maxmind providers is gated on Feature_XYZ, and Feature_XYZ is only available
```
Comment (jbaiera) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfiguration.java:141]: Can we update this comment with the right Feature name?

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseReaderLazyLoader.java [hunk 4]
```diff
@@ -277,4 +279,15 @@ private <RESPONSE> Optional<RESPONSE> lookup(Reader reader, String ip, Class<RES
             return Optional.of(builder.build(result, NetworkAddress.format(inetAddress), record.getNetwork(), List.of("en")));
         }
     }
+
+    long getBuilDateMillis() throws IOException {
```
Comment (jbaiera) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseReaderLazyLoader.java:283]: typo: getBuilDateMillis -> getBuildDateMillis

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportGetDatabaseConfigurationAction.java [hunk 5]
```diff
@@ -97,26 +118,76 @@ protected List<DatabaseConfigurationMetadata> createActionContext(Task task, Get
...
+            DatabaseConfigurationMetadata meta = geoIpMeta.getDatabases().get(id);
+            if (meta != null) {
+                maxmindDatabases.add(meta);
+            }
         }
-        return results;
+        return maxmindDatabases;
     }
 
     protected void newResponseAsync(
```
Comment (jbaiera) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportGetDatabaseConfigurationAction.java:193]: Maybe an oversight from previous work, but can we add @Override to this method?

**Review by jbaiera — APPROVED** [review]
LGTM! Only had a few small blurbs.

**Pull request merged**